knowledge_repo/app/models.py

```diff
@@ -234,26 +235,59 @@ class Vote(db.Model):
 
 
 @unique_constructor(
-    lambda username: username,
-    lambda query, username: query.filter(User.username == username)
+    lambda identifier: identifier,
+    lambda query, identifier: query.filter(User.identifier == identifier)
 )
-class User(db.Model):
+class User(db.Model, UserMixin):
     __tablename__ = 'users'
 
     id = db.Column(db.Integer, primary_key=True)
-    username = db.Column(db.String(500))
     created_at = db.Column(db.DateTime, default=func.now())
 
+    identifier = db.Column(db.String(500))  # Unique identifier across all login methods
+
+    username = db.Column(db.String(500))  # Username used to log in (may differ from identifier)
+    password = db.Column(db.String(500))  # Password for local logins
+
+    name = db.Column(db.String(500))  # Name as determined by auth method
+    preferred_name = db.Column(db.String(500))  # Name as determined by user preferences
+
+    email = db.Column(db.String(500))  # Email address
+    avatar_uri = db.Column(db.Text())  # Either external url or data uri
+    active = db.Column(db.Boolean, default=True)
+
+    last_login_at = db.Column(db.DateTime)  # Date of last login
+
     _posts_assoc = db.relationship("PostAuthorAssoc")
     posts = association_proxy('_posts_assoc', 'post')  # This property should not directly modified
 
+    # Method overrides for the UserMixin class for flask_login
+    @property
+    def is_active(self):
+        return self.active
+
+    @property
+    def is_authenticated(self):
+        return True
+
+    @property
+    def is_anonymous(self):
+        return False
+
+    def get_id(self):
+        return self.identifier
+
+    can_logout = True
+
+    # Other useful methods
     @property
     def format_name(self):
-        username_to_name = current_repo.config.username_to_name
-        return username_to_name(self.username)
+        return self.preferred_name or self.name or self.identifier
+        # username_to_name = current_repo.config.username_to_name
```

Comment:
The configuration for the repository and the configuration for the server interface do not **have** to line up, since the server interface can get a lot more information about users than is possible in the repository itself.

So far, we've lifted the configuration for names in the server directly from the primary repository, but this was actually only ever a stop-gap for this more complete authentication system... which can now get *actual* names as verified by various authentication providers. I think repositories will still need an `identifier_to_name` method because they are not privy to insight from the authentication providers.

I will do the updating of the repository side of things in a different PR, since this one is already huge; and delete that commented out code.